### DISCUSSION TITLE
MessagesPlaceholder in "Create New Prompt" panel

### DISCUSSION THREAD

**Discussion by NIKsun** [langfuse/langfuse#2210] [Ideas]
### Describe the feature or potential improvement

Hi, I've bumped into a little lack of functionality when trying to use Langfuse UI for prompt management. 

## Issue
I want to store a prompt that could be described in Langchain terms like:
```
prompt = ChatPromptTemplate.from_messages([
    ("system", "You are a helpful assistant."),
    ("placeholder", "{chat_history}"),
    ("user", "{input}"),
    ("placeholder", "{agent_scratchpad}"),
])
```
But I can't choose a placeholder "role" in the creation prompt panel:
![image](https://github.com/langfuse/langfuse/assets/13137917/eb9afb34-5f9d-4862-b7de-16023b035099)

## Potential solution
Extend the role selector with the "placeholder" option.

**Comment by marcklingen** [comment]
thanks for raising this. This is a mostly langchain specific change, we could generalize it to {chat_history} across our prompt management integrations. Currently you can achieve this by manually adding the history within the returned chat message array at runtime.

**Reply by mrdrprofuroboros** [reply]
Actually there's no validation, so creating such placeholder prompt from langfuse API works
but once you try to edit it in the UI it just shows unusable and blank

btw placeholder is a great feature, it is not only for "chat_history", you can have multiple placeholders in langchain and it is really convenient and way cleaner than adding in runtime

LangSmith supports placeholders :)

**Reply by marcklingen** [reply]
thanks for sharing!

**Comment by ericfeunekes** [comment]
I would suggest something like a named placeholder in the prompt template, which you could put anywhere. Then in the prompt compilation you would pass `<some_name>=[<list of messages]` and the list of messages would be expanded in place. You could assume it is always a list of messages since a single message is served by existing functionality. 

The named placeholder would allow for custom functionality like message history and likely other things that I can't think of right now.

**Reply by marcklingen** [reply]
thanks for sharing! agree on your take of how to implement this

**Reply by sfarthin** [reply]
A "messages" placeholder would be a valuable enhancement. This would clarify whether a prompt is expected to utilize an existing conversation chain or function independently. By making these expectations explicit within Langfuse, we can more easily understand the intended application flow without needing to refer to the code.

**Comment by debkanchan** [comment]
+1 I need this. Currently the workaround I use if using prompt template for system message only then _inalying_ inside `ChatPromptTemplate.fromMessages` like this:

```ts
ChatPromptTemplate.fromMessages([
  SystemMessagePromptTemplate.fromTemplate(prompt.getLangchainPrompt()),
  new MessagesPlaceholder("messages"),
])
```

One catch here is that (that's also implicitly mentioned in the original post) is that the placeholder is able to take any arbitrary key and not just `messages` key as in langgraph it is common to inlay multiple chat histories using different keys like this:

```ts
ChatPromptTemplate.fromMessages([
  SystemMessagePromptTemplate.fromTemplate(prompt.getLangchainPrompt()),
  new MessagesPlaceholder("messages"),
  new MessagesPlaceholder("agent_scratchpad")
])
```

**Comment by fbublitz** [comment]
+1 would be much appreciated here too, to have a chat history placeholder for messages in the prompt management as well as in the dataset experiments. I wouldnt say that it is such a langchain specific thing. As far as i know several LLM Api providers work with arrays of messages as input, so having the chat history as proper messages and not crammed into the system prompt or wrapped into a individual message would be very nice. 
We are also using the approach of @debkanchan, which is ok for running the chain. But for prompt experiments with datasets we would have to create another version of the original prompt that approximates the actual input to the llm and cannot be used afterwards in the chain.

**Comment by nimarb** [comment]
Hi @NIKsun , @mrdrprofuroboros , @ericfeunekes , @sfarthin , @debkanchan @fbublitz thank you all for the input. We have just released the Chat Message Placeholder feature, see the [changelog entry here](https://langfuse.com/changelog/2025-07-02-prompt-management-placeholders).

As per your input as well, a message placeholder has a slightly different structure:
```json
{
  "type": "placeholder",
  "name": "PLACEHOLDER_NAME"
}
```
where the `name` key can then be used to resolve the placeholder with a ChatMessage array (or anything really, if desired).

Would be happy about feedback / thoughts on it! You can try it in the playground and experiments.
Best, nimar
